### PR TITLE
[Kernel Selection] Remove hard code of PADDLE_WITH_MKLDNN (Part2 add dnn_fallback flag)

### PR DESCRIPTION
**Pull request by jiahy0825**
### PR types
Function optimization

### PR changes
OPs

### Describe
Delete the hardcodes of `#ifdef PADDLE_WITH_MKLDNN` and move such logic outside `GetExpectedKernelType`.

`OperatorWithKernel` adds the `dnn_fallback` flag, which determines whether fallback to plain kernel after calling  `GetExpectedKernelType`, use this bool flag to solve mkldnn and cudnn hardcode.

Relevant ops: `cast`, `transfer_dtype`, `layer_norm`, `pool2d`, `pool2d_grad`, `sum`, `sgd`, `reduce_max`, `reduce_mean`, `reduce_mean_grad`, `reduce_min`, `reduce_sum`, `reduce_sum_grad`.

Pre-PR: [[Kernel Selection] Remove hard code of PADDLE_WITH_MKLDNN #46606](https://github.com/PaddlePaddle/Paddle/pull/46606)

### CODE REVIEW

File: paddle/fluid/operators/optimizers/sgd_op.cc [hunk 1]
```diff
@@ -35,25 +35,20 @@ class SGDOp : public framework::OperatorWithKernel {
...
+    const auto *param_var = ctx.InputVar("Param");
+    const auto *grad_var = ctx.InputVar("Grad");
+
+    // supported cases
+    bool dense_param_sparse_grad = param_var->IsType<phi::DenseTensor>() &&
+                                   grad_var->IsType<phi::SelectedRows>();
+    bool dense_param_and_grad = param_var->IsType<phi::DenseTensor>() &&
+                                grad_var->IsType<phi::DenseTensor>();
+    if (!(dense_param_sparse_grad || dense_param_and_grad)) {
+      dnn_fallback_ = true;
```
Comment (chenwhql) [inline paddle/fluid/operators/optimizers/sgd_op.cc]: Maybe adding a `set_dnn_fallback` method is better? Changing the member of base class in method of child class is not a good impl

**Review by YuanRisheng — DISMISSED** [review]

File: paddle/fluid/operators/optimizers/sgd_op.cc (hunk 1 above)
Comment (jiahy0825) [reply paddle/fluid/operators/optimizers/sgd_op.cc]: Done, thanks~

**Review by zyfncg — APPROVED** [review]

**Review by chenwhql — APPROVED** [review]

**Comment by jiahy0825** [comment]
@Silv3S @jczaja Could you help me review this PR please?

**Review by Silv3S — APPROVED** [review]
LGTM

**Pull request merged**